[WIP] Make Inference/TrainingAgents accept torch tensors

ryotatomioka opened the pull request:
**Description**: This PR is an attempt to make the separation of responsibilities between Inference/TrainingManagers and Inference/TrainingAgents clearer.

In the proposed change, Inference/TrainingManagers are responsible for mapping the torch module parameters, buffers, inputs and outputs to onnx model initializers, inputs and outputs. However it does not perform any data type conversion.

On the other hand, Inference/TrainingAgents are responsible for converting data type between torch tensors and ORTValues.

This PR removes the possibility to pass a path to an onnx file when we construct Inference/TrainingAgents but I believe that is minor because it is easy to load the file first and then pass the model to the constructor. Other than that the PR should not change any behavior.

**Motivation and Context**
- Why is this change required? What problem does it solve? This change allows users to wrap arbitrary ONNX model in Inference/TrainingAgents. For example,
```python
import onnx
import torch
model = onnx.load_model("my_model.onnx")
device = torch.device("cpu")
agent = InferenceAgent(model, device)
inputs = [torch.randn(16, 784)]
outputs, run_info = agent.forward(*inputs)
```
- If it fixes an open issue, please link to the issue here.

thiagocrepaldi commented:
@ryotatomioka How does this PR interfere with https://github.com/microsoft/onnxruntime/pull/7691? Supporting ONNX model as input will be an experimental feature

thiagocrepaldi reviewed (commented):
@ryotatomioka How does this PR interfere with #7691? Supporting ONNX model as input will be an experimental feature

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,27 +2,40 @@
+class ExecutionAgent(object):
+    def __init__(self, onnx_model, device):
+        """
+        :param onnx_model: ONNX ModelProto
```
ORTModule uses google code style for python code

ryotatomioka replied:
Is the new version better?

ryotatomioka replied:
@thiagocrepaldi shall I also remove the `:param` annotation?

ryotatomioka replied:
done

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -109,27 +159,103 @@ def __init__(self, path_or_bytes, fw_feed_names, fw_outputs_device_info,
-        """
-        self._training_agent.run_backward(feeds, fetches, state)
+
+    def _prepare_backward_inputs(self, ctx, grad_outputs):
```
```suggestion
    def _process_backward_inputs(self, ctx, grad_outputs):
```
Better to make terminology uniform

ryotatomioka replied:
Thank you. Done

ryotatomioka commented:
@thiagocrepaldi If I understand correctly #7691 is a bigger change that involves a new feature to create an ORTModule from an ONNX model and perform AD on it.

This PR is a refactoring with no change to the high-level API. It just makes the Inference/TrainingAgent more accessible from pytorch.

I think the two PRs are combinable.

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
         self.run_id = run_id
 
+class ExecutionAgent(object):
+    """ExecutionAgent wraps ORT inference session and provides forward / backward methods
```
```suggestion
    """Executes the ONNX graph for both forward and backward calls
```

The use of InferenceSession is an implementation detail

baijumeswani commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+        #   especially the backward graph outputs.
+        # REVIEW(codemzs): Consolidate Training Agent with InferenceAgent on C++ side to not
+        # have the need for passing IOBinding.
+        io_binding = IOBinding(self._inference_session)
```
May we us ```self.io_binding()``` here so that we have a common place for creation of IOBinding?

baijumeswani commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+from onnxruntime.capi.onnxruntime_inference_collection import IOBinding, OrtValue, get_ort_device_type
+from onnxruntime.capi._pybind_state import OrtValueVector, PartialGraphExecutionState, RunOptions, TrainingAgent as C_TrainingAgent
+from . import _utils
+from ._graph_execution_manager import RunStateInfo
```
Does this not introduce a cyclic dependency? May we move ```RunStateInfo``` to a file of its own, or utils or here?

ryotatomioka replied:
No `_graph_execution_manager.py ` does not import this file

baijumeswani commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+
+@dataclass(frozen=True)
+class YieldOpInfo:
+    """Minimum version of GraphInfo needed for TrainingAgent
```
May we provide a more detailed docstring on what this class does?

ryotatomioka replied:
It doesn't do much. It is just to keep the utility function close to the values it returns. Any suggestion to improve the clarity is welcome.

thiagocrepaldi replied:
and why it is needed, since this class was introduced by this PR

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
-
-        self._inference_session = onnxruntime.InferenceSession(path_or_bytes, session_options,
+        super().__init__(onnx_model, device)
+        self._info = YieldOpInfo.from_training_model(onnx_model)
```
add a comment on why this is needed

thiagocrepaldi replied:
```suggestion
        self._yield_op_info = YieldOpInfo.from_training_model(onnx_model)
```
_info is too vague :)

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
         are given in 'providers', 'provider_options' should not be used.
 
         The list of providers is ordered by precedence. For example ['CUDAExecutionProvider', 'CPUExecutionProvider']
         means execute a node using CUDAExecutionProvider if capable, otherwise execute using CPUExecutionProvider.
```
```suggestion
        means execute a node using CUDAExecutionProvider if capable, otherwise execute using CPUExecutionProvider.

       Raises:
            YieldOpNotFound: when the model doesn't have a YieldOp node        
```
something like that

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+    full_shape_outputs: List[int]
+
+    @staticmethod
+    def from_training_model(onnx_model: ModelProto) -> "YieldOpInfo":
```
```suggestion
    def from_training_model(onnx_model: ModelProto) -> YieldOpInfo:
```

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_training_manager.py:
```diff
@@ -174,27 +118,23 @@ def backward(ctx, *grad_outputs):
-                        results.append(_utils._torch_tensor_from_dl_pack(
-                            backward_outputs.dlpack_at(initializer_index),
-                            backward_outputs[initializer_index]))
+                        results.append(backward_outputs[initializer_index])
```
what is the reasoning for this change? AFAIK this code was meant to keep high performance

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_training_manager.py:
```diff
@@ -174,27 +118,23 @@ def backward(ctx, *grad_outputs):
-                        results.append(_utils._torch_tensor_from_dl_pack(
-                            backward_outputs.dlpack_at(require_grad_names_index),
-                            backward_outputs[require_grad_names_index]))
+                        results.append(backward_outputs[require_grad_names_index])
```
what is the reasoning for this change? AFAIK this code was meant to keep high performance

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+    pass
+
+@dataclass(frozen=True)
+class YieldOpInfo:
```
My understanding is that this is a refactoring that gathered attributes from the onnx graph related to the YieldOp op. The original code was at `_ORTModuleFunction.backward`. Is that correct?

ryotatomioka replied:
Yes

thiagocrepaldi commented on orttraining/orttraining/python/training/ortmodule/_execution_agent.py:
```diff
@@ -2,134 +2,327 @@
+    full_shape_outputs: List[int]
+
+    @staticmethod
+    def from_training_model(onnx_model: ModelProto) -> "YieldOpInfo":
```
```suggestion
    def from_training_model(onnx_model: ModelProto) -> YieldOpInfo:
```
typo?

ryotatomioka replied:
This is a forward declaration because at this point the interpreter does not understand `YieldOpInfo`

thiagocrepaldi replied:
Only now I realized this method is inside the referred class :)

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_ortmodule_execution_agents.py:
```diff
@@ -0,0 +1,112 @@
+    return helper.make_model(graph_def)
+
+
+def assert_tensors_allclose(*args, **kwargs):
```
https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/test/python/_test_helpers.py#L179 already has assert helpers. If more are needed, put there

ryotatomioka replied:
I can use the one in the helper.

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_ortmodule_execution_agents.py:
```diff
@@ -0,0 +1,112 @@
+        helper.make_tensor_value_info("y", TensorProto.FLOAT, [2, 3]),
+        helper.make_tensor_value_info("bias", TensorProto.FLOAT, [3]),
+    ]
+    mul_node = helper.make_node("Mul", ["x", "y"], ["mul"])
+    add_node = helper.make_node("Add", ["mul", "bias"], ["add"])
+    outputs = [helper.make_tensor_value_info("add", TensorProto.FLOAT, [2, 3])]
+    graph_def = helper.make_graph([mul_node, add_node], "graph", inputs, outputs)
+    return helper.make_model(graph_def)
+
+
+def make_training_model():
+    inputs = [
+        helper.make_tensor_value_info("x", TensorProto.FLOAT, [2, 3]),
+        helper.make_tensor_value_info("y", TensorProto.FLOAT, [2, 3]),
+        helper.make_tensor_value_info("bias", TensorProto.FLOAT, [3]),
+    ]
+    mul_node = helper.make_node("Mul", ["x", "y"], ["mul"])
+    add_node = helper.make_node("Add", ["mul", "bias"], ["add"])
+    yield_node = helper.make_node(
+        "YieldOp", ["add"], ["add_grad"], full_shape_outputs=[0], domain="com.microsoft"
+    )
+    x_grad_node = helper.make_node("Mul", ["add_grad", "y"], ["x_grad"])
+    y_grad_node = helper.make_node("Mul", ["x", "add_grad"], ["y_grad"])
+    axes_tensor = helper.make_tensor("axes", TensorProto.INT64, [1], [0])
+    bias_grad_node = helper.make_node(
+        "ReduceSum", ["add_grad", "axes"], ["bias_grad"], keepdims=0
+    )
+    outputs = [
+        helper.make_tensor_value_info("x_grad", TensorProto.FLOAT, [2, 3]),
+        helper.make_tensor_value_info("y_grad", TensorProto.FLOAT, [2, 3]),
+        helper.make_tensor_value_info("bias_grad", TensorProto.FLOAT, [3]),
+    ]
+    graph_def = helper.make_graph(
+        [mul_node, add_node, yield_node, x_grad_node, y_grad_node, bias_grad_node],
+        "graph",
+        inputs,
+        outputs,
+        initializer=[axes_tensor],
+    )
+    return helper.make_model(graph_def)
```
Maybe move to https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/test/python/_test_commons.py? Maybe wrap on a class with methods to return both model and data

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_ortmodule_execution_agents.py:
```diff
@@ -0,0 +1,112 @@
+    np.testing.assert_allclose(*args, **kwargs)
+
+
+def execution_agent_factory(cls):
```
Try to keep only tests here. some versions of pytest really dont like to find functions that are not test in the same file

ryotatomioka replied:
Which version of pytest are we using? As far as I know, functions that doesn't start with `test_` are not considered tests. We can also configure pytest to look for more specific names
https://docs.pytest.org/en/6.2.x/example/pythoncollection.html

thiagocrepaldi replied:
It varies, in some pipelines it can be as old as 2.8.7

ryotatomioka replied:
How about marking them as `@pytest.fixture` then there should be no confusion?

thiagocrepaldi reviewed (commented):
Thanks for the PR, Ryota! Improvements are always welcome
Overall the changes make sense. However, I have a couple observations that I would like to discuss further:

* Why is proposed that  the user has to load ONNX model beforehand as opposed to accept path or bytes as currently implemented? This is not necessarily a blocker, but if there is a subtleness regarding this change, we need to know :)

* Most of the code is a refactoring, which is OK. I also noticed some changes that could have performance implications (although it seems to run fine) and the reason for it is not clear to me. One example is the change from

```
results.append(_utils._torch_tensor_from_dl_pack(
backward_outputs.dlpack_at(require_grad_names_index),
backward_outputs[require_grad_names_index]))
```
to
```
results.append(backward_outputs[require_grad_names_index])
```
Could you elaborate on that?

* Could you enumerate the actual changes (not just refactoring) such as the above and their reasoning? We have spent quite a time making some tuning on the frontend to prevent memory copy, etc and we have to make sure we won't accidentally regress the perf despite the code still runs fine

* There is a pipeline failure that needs to be addressed

ytaous commented:
looks like the logic has been refactored to _execution_agent.backward(), thus should be equivalent

---
In reply to: [704509211](https://github.com/microsoft/onnxruntime/pull/8321#issuecomment-704509211) [](http://example.com/codeflow?ancestors=704509211,704509211)

ryotatomioka commented:
> * Most of the code is a refactoring, which is OK. I also noticed some changes that could have performance implications (although it seems to run fine) and the reason for it is not clear to me. One example is the change from
> 
> ```
> results.append(_utils._torch_tensor_from_dl_pack(
> backward_outputs.dlpack_at(require_grad_names_index),
> backward_outputs[require_grad_names_index]))
> ```
> 
> to
> 
> ```
> results.append(backward_outputs[require_grad_names_index])
> ```
> 
> Could you elaborate on that?

If I understand correctly (I have added asserts to check my understanding) what is happening here is a scatter. Every element of input `backward_outputs` appears in the output `results` although some elements of `results` may be empty (i.e. `None`).

If my above understanding is correct, there is no behavior change in my refactoring. I have just moved the `_utils._torch_tensor_from_dl_pack(...)` call into `TrainingAgent`.

thiagocrepaldi commented on orttraining/orttraining/test/python/orttraining_test_ortmodule_execution_agents.py:
```diff
@@ -1,18 +1,17 @@
 
 
-def make_inference_model():
+@pytest.fixture
```
did you test this trick on pytest 2.8.7 by running pytest -sv orttraining/orttraining/test/python/orttraining_test_ortmodule_execution_agents.py directly?

ryotatomioka replied:
I will try

ryotatomioka replied:
Yes it worked (on CPU)
```
================================================= test session starts ==================================================
platform linux -- Python 3.6.9, pytest-2.8.7, py-1.9.0, pluggy-0.3.1 -- /usr/bin/python
cachedir: ../../../.cache
rootdir: /home/ryoto/onnxruntime, inifile:
collecting 0 items2021-07-15 05:40:25.551692361 [E:onnxruntime:Default, provider_bridge_ort.cc:950 Ensure] Failed to load library libonnxruntime_providers_shared.so with error: libonnxruntime_providers_shared.so: cannot open shared object file: No such file or directory
2021-07-15 05:40:25.551752662 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:1688 CreatePybindStateModule] Init provider bridge failed.
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
collected 2 items

orttraining_test_ortmodule_execution_agents.py::test_onnx_graph_forward PASSED
orttraining_test_ortmodule_execution_agents.py::test_onnx_graph_forward_backward PASSED

=============================================== 2 passed in 1.43 seconds ===============================================
```

Is `Init provider bridge failed` something we should worry about?

thiagocrepaldi commented:
Temporarily on hold until the fallback design is completed

thiagocrepaldi commented:
@ryotatomioka can you try to resolve the conflicts for this PR?
